### PR TITLE
[SPARK-40539][CONNECT] Initial DataFrame Read API parity for Spark Connect

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Add initial Read API for Spark Connect that allows setting schema, format, option and path, and then to read files into DataFrame.

### Why are the changes needed?

PySpark readwriter API parity for Spark Connect

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
@grundprinzip @HyukjinKwon @cloud-fan

File: connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -81,7 +81,26 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
         } else {
           child
         }
-      case _ => throw InvalidPlanInput()
+      case proto.Read.ReadTypeCase.FILE_DATA_SOURCE =>
+        val reader = session.read
+        if (rel.getFileDataSource.getFormat != null) {
```
Comment (amaliujia) [inline connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:86]: I need to double check if there is a better way to check if a Java proto filed is not populated. I didn't find a `hasXXX` method in Intellij code auto-complete list.

Comment (amaliujia) [reply connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:86]: BTW I found that proto3 does not support `required` on proto fields anymore so we cannot rely on that.

File: connect/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -64,11 +64,25 @@ message SQL {
...
+
+  message FileDataSource {
+    repeated string path = 1;
+    string format = 2;
+    string schema = 3;
+    repeated DataSourceOptionPair options = 4;
+  }
+
+  message DataSourceOptionPair {
+    // option key is case insensitive.
```
Comment (grundprinzip) [inline connect/src/main/protobuf/spark/connect/relations.proto:82]:
```suggestion
    // Option key is case insensitive.
```

File: connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 1 above)
Comment (grundprinzip) [inline connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:86]:
```suggestion
        if (rel.getFileDataSource.hasFormat) {
```

File: connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -81,7 +81,26 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
...
+          reader.format(rel.getFileDataSource.getFormat)
+        }
+        if (rel.getFileDataSource.getSchema != null) {
```
Comment (grundprinzip) [inline connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:89]:
```suggestion
        if (rel.getFileDataSource.hasSchema) {
```

**Review by grundprinzip — COMMENTED** [review]
I left some suggestions on some of the pieces. Nothing major.

File: connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 3 above)
Comment (amaliujia) [reply connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:89]: This is the first I tried and I am pretty there is no `hasSchema`. This is so wried to me that the generate class does not have a `contains` like calls on these fields.

File: connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 4]
```diff
@@ -81,7 +81,26 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
...
+          reader.schema(rel.getFileDataSource.getSchema)
+        }
+        if (rel.getFileDataSource.getOptionsCount > 0) {
+          for (x <- 1 until rel.getFileDataSource.getOptionsCount) {
+            val optionPair = rel.getFileDataSource.getOptions(x)
+            reader.option(optionPair.getOptionKey, optionPair.getOptionValue)
+          }
+        }
+        if (rel.getFileDataSource.getPathList != null) {
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: There is no `hasPathList` but I choose to use `getPathCount > 0` to replace this `null` check.

File: python/pyspark/sql/connect/readwriter.py [hunk 5]
```diff
@@ -31,6 +39,104 @@ class DataFrameReader:
...
+        if format is not None:
+            self.format(format)
+        if schema is not None:
+            self.schema(schema)
+        for k in options.keys():
+            self.option(k, options.get(k))
+        if isinstance(path, str):
+            self._path.append(path)
+        elif path is not None:
+            if type(path) != list:
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/readwriter.py:132]:
```suggestion
            if not isinstance(path, list):
```

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 6]
```diff
@@ -65,11 +65,25 @@ message SQL {
...
+    repeated string path = 1;
+    string format = 2;
+    string schema = 3;
+    repeated DataSourceOptionPair options = 4;
+  }
+
+  message DataSourceOptionPair {
+    // Option key is case insensitive.
+    string option_key = 1;
+    string option_value = 2;
```
Comment (grundprinzip) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:85]: Sorry for the late reply, isn't the option_value actually more a literal?

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:85]: They are strings: https://github.com/apache/spark/blob/357193d55508051369228c5e7e252c17089da317/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L703

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 7]
```diff
@@ -65,11 +65,25 @@ message SQL {
 message Read {
   oneof read_type {
     NamedTable named_table = 1;
+    FileDataSource file_data_source = 2;
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:68]: Refactor proto

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 8]
```diff
@@ -65,11 +65,25 @@ message SQL {
...
   }
 
   message NamedTable {
     repeated string parts = 1;
   }
+
+  message FileDataSource {
+    repeated string path = 1;
+    string format = 2;
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:77]: required

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 9]
```diff
@@ -65,11 +65,25 @@ message SQL {
...
+    string schema = 3;
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:78]: optional

File: python/pyspark/sql/connect/plan.py [hunk 10]
```diff
@@ -101,6 +101,32 @@ def _child_repr_(self) -> str:
         return self._child._repr_html_() if self._child is not None else ""
 
 
+class FileDataSource(LogicalPlan):
+    def __init__(
+        self,
+        path: Optional[List[str]] = None,
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/plan.py:107]: `path` should end up with options.

File: python/pyspark/sql/connect/plan.py [hunk 11]
```diff
@@ -101,6 +101,32 @@ def _child_repr_(self) -> str:
...
+        format: Optional[str] = None,
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/plan.py:108]: `format` is required

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 12]
```diff
@@ -65,11 +65,25 @@ message SQL {
...
+    repeated DataSourceOptionPair options = 4;
```
Comment (zhengruifeng) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:79]: maybe simply `map<string, string>`?

File: python/pyspark/sql/connect/readwriter.py [hunk 13]
```diff
@@ -31,6 +39,104 @@ class DataFrameReader:
...
+
+        Parameters
+        ----------
+        key : str
+            The key for the option to set. key string is case-insensitive.
+        value
+            The value for the option to set.
+
+        """
+        self._options[key] = str(value)
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/readwriter.py:97]: do we need to convert `key` to lower case?

Comment (amaliujia) [reply python/pyspark/sql/connect/readwriter.py:97]: the key is case insensitive and the server side will convert it anyway.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 12 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:79]: oh yes.. I forget that proto supports `MAP`. Will update this.

**Comment by amaliujia** [comment]
@HyukjinKwon @zhengruifeng @cloud-fan 

This PR finally has caught up after blockers were solved. It is ready for another review now.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 14]
```diff
@@ -91,7 +92,25 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
...
+
+        val reader = session.read
+        reader.format(rel.getDataSource.getFormat)
+        for ((k, v) <- localMap) {
+          reader.option(k, v)
+        }
+        if (rel.getDataSource.getSchema != null) {
+          reader.schema(rel.getDataSource.getSchema)
+        }
+        reader.load().queryExecution.analyzed
```
Comment (grundprinzip) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:112]: So my question here is, if the execution has already happened or not. What I mean is, is `load()` a blocking operation or a logical one? If you call `load()` and then return the analyzed plan. What happens if you call collect on this plan? Does the load happen again?

Can verify this?

File: python/pyspark/sql/connect/plan.py [hunk 15]
```diff
@@ -101,6 +102,30 @@ def _child_repr_(self) -> str:
         return self._child._repr_html_() if self._child is not None else ""
 
 
+class DataSource(LogicalPlan):
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/plan.py:105]: Couple of nits:

- doc?
- repr / repr_html?

Thanks!

**Review by grundprinzip — COMMENTED** [review]
Please have a look at my main comment in `SparkConnectPlanner` with regard to calling `load()` during the planning.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 16]
```diff
@@ -108,7 +109,25 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
         } else {
           child
         }
-      case _ => throw InvalidPlanInput()
+      case proto.Read.ReadTypeCase.DATA_SOURCE =>
+        if (rel.getDataSource.getFormat == "") {
+          throw InvalidPlanInput("DataSource requires a format")
+        }
+        val localMap = CaseInsensitiveMap[String](rel.getDataSource.getOptionsMap.asScala.toMap)
+        if (!localMap.contains("path")) {
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:117]: no, data source does not require a path option.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 17]
```diff
@@ -91,7 +92,25 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
...
+          throw InvalidPlanInput("DataSource requires a path")
+        }
+
+        val reader = session.read
+        reader.format(rel.getDataSource.getFormat)
+        for ((k, v) <- localMap) {
+          reader.option(k, v)
+        }
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:125]: I like this shorter one line version

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 14 above)
Comment (cloud-fan) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:129]: `load()` just builds the logical plan. it's not an action.

File: python/pyspark/sql/connect/readwriter.py [hunk 18]
```diff
@@ -29,8 +37,114 @@ class DataFrameReader:
...
     """
 
-    def __init__(self, client: "RemoteSparkSession") -> None:
+    def __init__(self, client: "RemoteSparkSession"):
         self._client = client
+        self._format = ""
+        self._schema = ""
+        self._options: Dict[str, str] = {}
+
+    def format(self, source: str) -> "DataFrameReader":
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply python/pyspark/sql/connect/readwriter.py:46]: Yes these are to match DataFrame API.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 16 above)
Comment (amaliujia) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: I see. Removed

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 19]
```diff
@@ -65,11 +65,21 @@ message SQL {
...
     NamedTable named_table = 1;
+    DataSource data_source = 2;
   }
 
   message NamedTable {
     repeated string parts = 1;
   }
+
+  message DataSource {
+    // Required. Supported formats include: parquet, orc, text, json, parquet, csv, avro.
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:77]: I believe these formats are called `built-in` format and we can trust that Spark will always support those by default.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:77]: `jdbc` is also a built-in format. I think it's OK to just give some examples here.

File: python/pyspark/sql/connect/plan.py [hunk 20]
```diff
@@ -111,6 +112,46 @@ def _child_repr_(self) -> str:
...
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/plan.py:115]: is it generated by proto?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 19 above)
[the remark this answers is not in the document]
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:77]: I really like how Apache Beam document their proto and I want to match it in connect once the proto becomes stable. 

So this part will be expanded anyway (e.g. include the full list, document case sensitivity, document applicable options for each format if there is any, etc.)

File: python/pyspark/sql/connect/plan.py (hunk 20 above)
Comment (amaliujia) [reply python/pyspark/sql/connect/plan.py:115]: This is not. This class is more or less the same as the DSL that we introduced for Scala: the core idea is to provide a way for `toProto`.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 14 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:126]: Thanks I was able to verify this for myself as well using some experiments.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by HyukjinKwon — COMMENTED** [review]
late LGTM2

**Pull request closed without merging**